### ISSUE TITLE
FP S4050: Issue is raised when operator method implementations are provided through an interface

### ISSUE THREAD

**Issue by mary-georgiou-sonarsource** [SonarSource/sonar-dotnet#6442]
### Description

Since C# 11 there's the possibility to have static virtual method implementations in an interface.
For this rule when the operator implementations are provided through an interface an issue is still raised.

### Repro steps

```csharp
public class Foo : IEquality<Foo> // Noncompliant FP {{Provide an implementation for: 'operator==' and 'operator!='.}}
{
    public static Foo operator +(Foo a, Foo b) => new Foo();

    public override bool Equals(object obj) => false;
    public override int GetHashCode() => 0;
}

public interface IEquality<TSelf> where TSelf : IEquality<TSelf>
{
    static virtual bool operator ==(TSelf a, TSelf b) => true;
    static virtual bool operator !=(TSelf a, TSelf b) => false;
}
```

### Expected behavior

No issue should be raised

* C#/VB.NET Plugins version 8.49

**Comment by martin-strecker-sonarsource** [comment]
I think the issue is wrong. The assumption that the operators are provided through the interface is not entirely true. See e.g.
```cs
using System;

var foo1 = new Foo();
var foo2 = new Foo();
Console.WriteLine($"Via concrete instances: {foo1 == foo2}");
M(foo1, foo2);

void M<T>(T t1, T t2) where T:IEquality<T>
{
  Console.WriteLine($"Via constrained generic parameters: {t1 == t2}");
}

public class Foo : IEquality<Foo> // Noncompliant FP {{Provide an implementation for: 'operator==' and 'operator!='.}}
{
    public static Foo operator +(Foo a, Foo b) => new Foo();

    public override bool Equals(object obj) => false;
    public override int GetHashCode() => 0;
}

public interface IEquality<TSelf> where TSelf : IEquality<TSelf>
{
    static virtual bool operator ==(TSelf a, TSelf b) => true;
    static virtual bool operator !=(TSelf a, TSelf b) => false;
}
```
This prints out (see it [here](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgNwEMwACAMwHtLNSBeUqAUwHdSAxagCgEoBuQiQrV09Rqw7d+eVJgCcXACQAiAGoBLIqQDGlKNrBNgTUuqgBnYEX1NzIUgG8qNeg2foAvsum4Asl2dMGGFKdB8ZABZSXwAeABUAPi440mAg0hTgMNIWAAsmQwyQAEkAUQBHCCIAG3VgAE94hLwHPFJSWQUVDS1dC2AwIjMmOFIAcyZmMHVtUgAHEiIAWyMCu0c011TPbwFcDzwZAGYO0U5KUnsyypq6xvOE0gB6J9IAOT1dJbna62AOAAKjgcALAlAI6jgJmspm+1SYKygVmA6j0wjA9gA5JQ5gUiMBKGA6HRMaRrKNsbjBgSwABCEkAOg8B1wrVw7XaqBOsgAbJILji8TTSIguOcycFxQAjHj0R7MNjnXh7NocrmkcEFaZQ0hS6jVUgVKrVcxcShSgBWTG0/3NFtldEe5Bq5iYew5HROmrA2pMZn+AHEjAAJIjmXIAYUoUN4ctIAAY9izjqYkQVndoTNdjXd4gBlJjVciPPIFExxAtFy6kbO3Br5wvFlqqzmYPkQsDAY26/Uaqn4wmuZKV8gSjIj3UOx4DCBulsdNukDtdmo9ygGwXUwf04eNscVvcyuPOk1z/Z4IA===) yourself:
```
Via concrete instances: False
Via constrained generic parameters: True
```
So in the first sample, the built-in equality operators are chosen while in the second sample the operators defined by the interface are chosen.

I therefore would not consider this an FP.

**Comment by mary-georgiou-sonarsource** [comment]
Closing - this is not a valid FP.
